### PR TITLE
[Diffusion][Refactor] Refactor and extract complex RoPE implementation to layers/rotary_embedding for MOVA DiT

### PR DESCRIPTION
**Pull request by Allor-maker**
## Motivation

SGLang has a dedicated directory (`layers/rotary_embedding`) designed to house unified, high-performance Rotary Positional Embedding implementations. 
 
Was observed that a complex-valued RoPE representation is highly efficient for the `neox_style=False` and `interleaved=True` configurations. However, implementing this complex multiplication logic directly within model-specific code leads to duplication and limits its reusability. 
 
Since this optimized complex RoPE path is generic, mathematically clean, and performant, it is logical to extract the implementation out of the model code and move it into the shared `layers/rotary_embedding/utils.py` utility module. This enables other models within SGLang using interleaved RoPE to easily inherit and reuse this optimization.

## Modifications

Were modified 4 key files to implement this refactoring and optimization:
 
### 1. `sglang/multimodal_gen/runtime/layers/rotary_embedding/utils.py`
- Added the `_apply_rotary_emb_complex` function:
  - **Inputs:** 
    - `x` of shape `[batch, num_tokens, num_heads, head_size]`.
    - `freqs` of shape `[num_tokens, 1, head_size // 2]`.
  - **Outputs:** Evaluated tensor of the same shape and dtype as `x`.
  - **Optimization Details:** It utilizes `view()` transformations to group activation dimensions and cast to `float32`. It then performs a single vectorized complex multiplication kernel. Calling `torch.view_as_real()` on the result natively lays out the real and imaginary elements sequentially in memory, producing the interleaved layout. This excludes the need for `torch.stack()` and `.contiguous()` allocations.
 
### 2. `sglang/multimodal_gen/runtime/layers/rotary_embedding/__init__.py`
- Registered the new `_apply_rotary_emb_complex` utility in the package exports: imported the function from `.utils.py` and added it to the `__all__` list to expose it cleanly to the rest of the codebase.
 
### 3. `sglang/multimodal_gen/runtime/models/dits/mova_video_dit.py` and `mova_audio_dit.py`
Cleaned up the attention blocks and models:
  - Removed custom inline complex RoPE functions.
  - In `SelfAttention.forward`, reshape `q`, `k`, and `v` to `[B, S, H, D]` format using `view()` operations, and apply the `_apply_rotary_emb_complex` directly. This eliminates `rearrange` and `flatten` steps inside the attention loop.
 

## Accuracy Tests

https://github.com/user-attachments/assets/9239f012-ea90-496d-8823-01fe6d95584e

### 1. SGLang Server Launch Command
```bash
sglang serve --model-path $(pwd)/mova_video_360p --pipeline-class-name MOVAPipeline --num-gpus 4 --ring-degree 1 --ulysses-degree 4 --performance-mode memory --dit-layerwise-offload --dit-offload-prefetch-size 0.1 --text-encoder-cpu-offload true --vae-cpu-offload true --enable-torch-compile --warmup-mode server --vae-tiling --vae-sp
```

### 2. Video Generation Request

```bash
curl -X POST "http://localhost:30000/v1/videos" -F "prompt=A curious raccoon" -F "input_reference=@/.../images/raccoon.png" -F "size=640x352"  -F "num_frames=49" -F "fps=24" -F "num_inference_steps=25" -F "guidance_scale=5.0" -o create_video.json
```

### 3. Offline Profiling Command
```bash
SGLANG_DIFFUSION_TORCH_PROFILER_DIR="/home/s60128115/profile_complex" sglang generate --model-path /home/s60128115/models/mova_video_360p --pipeline-class-name MOVAPipeline --prompt "A curious raccoon" --image-path /home/s60128115/images/raccoon.png --width 640 --height 352 --num-frames 49 --fps 24 --num-inference-steps 25 --guidance-scale 5.0 --profile --num-profiled-timesteps 10 --num-gpus 4 --ulysses-degree 4 --enable-torch-compile --warmup-mode request --dit-layerwise-offload --dit-offload-prefetch-size 0.1
```

### 4. Profiler Latency Results (Ascend NPU)
The table below compares the detailed execution times of the rope functions on NPU before and after modifications:
 
| Implementation | func_name | time(%) | total_time(us) | num_calls | Avg(us) | Min(us) | Max(us) |
| :--- | :---: | :---: | :---: | :---: | :---: | :---: | :---: |
| **Baseline** | rope_apply_head_dim | 100 | 	8983898.08 | 3080 | 2916.85 | 2378.38 | 7904.13 |
| **Modified** | _apply_rotary_emb_complex | 100 | 7139799.61 | 3080 | 2318.1168 | 2021.59 | 5624.96 |

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :x: [Run #32064673694](https://github.com/sgl-project/sglang/actions/runs/32064673694)
Latest PR Test (Extra): :x: [Run #32064673343](https://github.com/sgl-project/sglang/actions/runs/32064673343)

### CODE REVIEW

**Review by Makcum888e — APPROVED** [review]
LGTM

**Comment by ping1jing2** [comment]
> Since this optimized complex RoPE path is generic, mathematically clean, and performant, it is logical to extract the implementation out of the model code and move it into the shared layers/rotary_embedding/utils.py utility module. This enables other models within SGLang using interleaved RoPE to easily inherit and reuse this optimization.

do you test other models with complex RoPE, and can we get performance gain from it?

> 4. Profiler Latency Results (Ascend NPU)

can you explain why we can get performance gain if we just move such RoPE from model to common dir?

**Comment by Allor-maker** [comment]
Thank you for the review!
1. I also tested the QwenImage model, and the `_apply_rotary_emb_complex` function allowed for improved performance, as seen in the table. This function is relevant for models using the RoPE parameter with `is_neox_style=False` on NPU, as this implementation is faster than the current fallback for this type of device and argument. I am currently preparing a PR with a class that unifies the call to RoPE, which will be used in subsequent PRs to analyze performance on other DIT models.

| Implementation | func_name | time(%) | total_time(us) | num_calls | Avg(us) | Min(us) | Max(us) |
| :--- | :---: | :---: | :---: | :---: | :---: | :---: | :---: |
| **Baseline** | apply_rotary_embedding_native | 100 | 13628501.78 | 2448 | 5567.1984 | 4108.45 | 12372.26 |
| **Modified** | _apply_rotary_emb_complex | 100 | 9655076.15 | 2448 | 3944.0671 | 2750.82 | 8124 |
2. In the current implementation, the calls to `rearrange` and `flatten` have been replaced with `view`, which slightly improves performance.

**Review by ping1jing2 — APPROVED** [review]
